src/main/java/com/exadel/discount/filters/JwtFilter.java

```diff
@@ -0,0 +1,105 @@
+
+    private final UserDetailsServiceImpl userDetailsService;
+
+    private AbstractAuthenticationToken authentication;
+    private String token;
+    private String username;
+    private String role;
+
+    @Override
+    protected void doFilterInternal(HttpServletRequest request,
+                                    HttpServletResponse response,
+                                    FilterChain filterChain) throws ServletException, IOException {
+        final String authorizationHeader = request.getHeader(AUTHORIZATION_HEADER);
+
+        if (isNotAuthorized() && hasSuitableAuthorizationHeader(authorizationHeader)) {
+            extractPayload(authorizationHeader);
+
+            if (hasValidSearchKeyForStorage()) {
+                UserDetails userDetails = loadUserDataFromStorageThroughSearchKey();
+
+                if (StringUtils.equals(role, JwtVariablesConfig.getProperty(REFRESH_ROLE))) {
+                    if (JwtUtil.validateToken(token, userDetails)) {
+                        this.authentication = new PreAuthenticatedAuthenticationToken(
+                                userDetails, null, RefreshJwtUtil.getRefreshAuthority());
+                    }
+                } else if (JwtUtil.validateToken(token, userDetails)) {
+                    this.authentication = new UsernamePasswordAuthenticationToken(
+                            userDetails, null, userDetails.getAuthorities());
+
+                }
+                setSecurityContextAuthentication(request, authentication);
+            }
+        }
+        filterChain.doFilter(request, response);
+    }
+
+    private boolean isNotAuthorized() {
+        return SecurityContextHolder.getContext().getAuthentication() == null;
+    }
+
+    private boolean hasSuitableAuthorizationHeader(String authorizationHeader) {
+        return
+                StringUtils.isNoneEmpty(authorizationHeader) &&
+                        StringUtils.startsWith(authorizationHeader, BEARER_TYPE_OF_AUTHORIZATION_HEADER);
+    }
+
+    private UserDetails loadUserDataFromStorageThroughSearchKey() {
+        return this.userDetailsService.loadUserByUsername(username);
+    }
+
+    private boolean hasValidSearchKeyForStorage() {
+        return StringUtils.isNoneEmpty(username);
+    }
+
+    private void setSecurityContextAuthentication(HttpServletRequest request, AbstractAuthenticationToken authentication) {
+        authentication.setDetails(new WebAuthenticationDetailsSource().buildDetails(request));
+        SecurityContextHolder.getContext().setAuthentication(authentication);
+    }
+
+    private void extractPayload(String authorizationHeader) {
+        setToken(authorizationHeader.substring(7));
+        setUsername(JwtUtil.extractUsername(token));
+        setRole(JwtUtil.extractRole(token));
+        extractRoleKeyWordFromClaim();
+    }
+
+    private void extractRoleKeyWordFromClaim() {
+        final String preSyntax = "[{authority=";
```

Comment:
Why role come in such ugly way? It should be string of array of strings in tocken